blog: Windows Operational Readiness (#45130)
kubernetes/website: content/en/blog/_posts/2024-04-03-windows-ops-readiness.md

Since Windows support [graduated to stable](/blog/2019/03/25/kubernetes-1-14-release-announcement/)
with Kubernetes 1.14 in 2019, the capability to run Windows workloads has been much
appreciated by the end user community. The level of and availability of Windows workload
support has consistently been a major differentiator for Kubernetes distributions used by
large enterprises. However, with more Windows workloads being migrated to Kubernetes
and new Windows features being continuously released, it became challenging to test
Windows worker nodes in an effective and standardized way.

The Kubernetes project values the ability to certify conformance without requiring a 
closed-source license for a certified distribution or service that has no intention 
of offering Windows.

Some notable examples brought to the attention of SIG Windows were:

- An issue with load balancer source address ranges functionality not operating correctly on
  Windows nodes, detailed in a GitHub issue:
  [kubernetes/kubernetes#120033](https://github.com/kubernetes/kubernetes/issues/120033).
- Reports of functionality issues with Windows features, such as

>   “[GMSA](https://learn.microsoft.com/en-us/windows-server/security/group-managed-service-accounts/group-managed-service-accounts-overview) isn’t working, why?”,

Reviewer: The previous sentence didn't read well and wasn't the issue's title. I've suggested an alternative phrasing below.
```suggestion
  “[GMSA](https://learn.microsoft.com/en-us/windows-server/security/group-managed-service-accounts/group-managed-service-accounts-overview) not working with containerd”,
```

which are standardized tests designed to ensure that a Kubernetes cluster meets 
the required Kubernetes specifications.

However, these tests were originally defined at a time when Linux was the *only* 
operating system compatible with Kubernetes, and thus, they were not easily 
extendable for use with Windows. Given that Windows workloads, despite their 
importance, account for a smaller portion of the Kubernetes community, it was 
important to ensure that the primary conformance suite relied upon by many 
Kubernetes distributions to certify Linux conformance, didn't become encumbered 
with Windows specific features or enhancements such as GMSA or multi-operating 
system kube-proxy behavior.

Therefore, since there was a specialized need for Windows conformance testing, 
SIG Windows went down the path of offering Windows specific conformance tests 
through the Windows Operational Readiness Specification.

## Can’t we just run the Kubernetes end-to-end test suite?

In the Linux world, tools such as [Sonobuoy](https://sonobuoy.io/) simplify execution of the 
conformance suite, relieving users from needing to be aware of Kubernetes' 

> compilation paths or the semantics of Ginkgo tags.

Reviewer: ```suggestion
compilation paths or the semantics of [Ginkgo tags](https://onsi.github.io/ginkgo).
```

- An issue with load balancer source address ranges functionality not operating correctly on
  Windows nodes, detailed in a GitHub issue:
  [kubernetes/kubernetes#120033](https://github.com/kubernetes/kubernetes/issues/120033).
- Reports of functionality issues with Windows features, such as
  “[GMSA](https://learn.microsoft.com/en-us/windows-server/security/group-managed-service-accounts/group-managed-service-accounts-overview) isn’t working, why?”,
  discussed in [microsoft/Windows-Containers#44](https://github.com/microsoft/Windows-Containers/issues/44).
- Challenges developing networking policy tests that could objectively evaluate
  Container Network Interface (CNI) plugins across different operating system configurations,
  as discussed in [kubernetes/kubernetes#97751](https://github.com/kubernetes/kubernetes/issues/97751).

SIG Windows therefore recognized the need for a tailored solution to ensure Windows
nodes' operational readiness *before* their deployment into production environments.
Thus, the idea to develop a [Windows Operational Readiness Specification](https://kep.k8s.io/2578)
was born.

## Can’t we just run the official Conformance tests?

> The Kubernetes project contains a set of [conformance tests](https://github.com/kubernetes/kubernetes/blob/master/test/conformance/testdata/conformance.yaml), 

Reviewer 2: Let's link instead to one of:
- https://www.cncf.io/training/certification/software-conformance/#how
- https://kubernetes.io/docs/setup/best-practices/node-conformance/

which are standardized tests designed to ensure that a Kubernetes cluster meets 
the required Kubernetes specifications.

> 

Reviewer 2: Optionally, mention the existing (Linux-only) node conformance tests, https://github.com/kubernetes/community/blob/master/contributors/devel/sig-node/e2e-node-tests.md

Reviewer 3: Leaving this outside the blog because there's no direct relation with the post.

However, these tests were originally defined at a time when Linux was the *only* 
operating system compatible with Kubernetes, and thus, they were not easily 

> extendible for use with Windows. Given that Windows workloads, despite their 

Reviewer 2: Try “extendable” or “extensible”.

system kube-proxy behavior.

Therefore, since there was a specialized need for Windows conformance testing, 
SIG Windows went down the path of offering Windows specific conformance tests 
through the Windows Operational Readiness Project.

## Can’t we just run the Kubernetes end-to-end test suite?

In the Linux world, tools such as [Sonobuoy](https://sonobuoy.io/) simplify execution of the 
conformance suite, relieving users from needing to be aware of Kubernetes' 
compilation paths or the semantics of Ginkgo tags.

Regarding needing to compile the Kubernetes tests, we realized that Windows 
users might similarly find the process of compiling and running the Kubernetes 
e2e suite from scratch similarly undesirable, hence, there was a clear need to 
provide a user-friendly, "push-button" solution that is ready to go. Moreover, 
regarding Ginkgo tags, applying conformance tests to Windows nodes through a set 
of [Ginkgo](https://onsi.github.io/ginkgo/) tags would also be burdensome for 
any user, including Linux enthusiasts or experienced Windows system admins alike.

To bridge the gap and give users a straightforward way to confirm their clusters 

> support a variety of features, the SIG Windows community found it necessary to 

Reviewer 2: ```suggestion
support a variety of features, the Kubernetes SIG for Windows found it necessary to
```

| `Extend.Network`         | Tests advanced networking functionality, (ability to support IPv6)                                                                  |
| `Extend.Worker`          | Tests features related to Windows worker node functionality, (ability for nodes to access TCP and UDP services in the same cluster) |

## How to conduct operational readiness tests for Windows nodes

To run the Windows Operational Readiness test suite, refer to the test suite's
[`README`](https://github.com/kubernetes-sigs/windows-operational-readiness/blob/main/README.md), which explains how to set it up and run it. The test suite offers 
flexibility in how you can execute tests, either using a compiled binary or a 
Sonobuoy plugin. You also have the choice to run the tests against the entire 
test suite or by specifying a list of categories. Cloud providers have the 
choice of uploading their conformance results, enhancing transparency and reliability.

Once you have checked out that code, you can run a test. For example, this sample 
command runs the tests from the `Core.Concurrent` category:

```shell

> $ ./op-readiness --kubeconfig $KUBE_CONFIG --category Core.Concurrent

Reviewer 2: We [don't include the prompt](https://kubernetes.io/docs/contribute/style/style-guide/#don-t-include-the-command-prompt), so that the blocks can be valid code snippets
```suggestion
./op-readiness --kubeconfig $KUBE_CONFIG --category Core.Concurrent
```

```

> If you want to test your changes against a specific pull request using the 

Reviewer 2: ```suggestion
As a contributor to Kubernetes, if you want to test your changes against a specific pull request using the 
```

Windows Operational Readiness Specification, use the following bot command in 
the new pull request.

```
/test operational-tests-capz-windows-2019
```

## Looking ahead

We’re looking to improve our curated list of Windows-specific tests by adding 
new tests to the Kubernetes repository and also identifying existing test cases 
that can be targetted. The long term goal for this subproject is to continually 
enhance test coverage for Windows worker nodes and improve the robustness of 
Windows support, facilitating a seamless experience across diverse cloud 
environments. We also have plans to integrate the Windows Operational Readiness 
tests into the official Kubernetes conformance suite.

> 

Reviewer 2: Is there any plan for a Linux Operational Readiness Specification?

Reviewer 3: No plans for a Linux op readiness. The idea is to evolve this specification to be an official "conformance Windows test suite" in the future.

If you are interested in helping us out, please reach out to us! We welcome help 
in any form, from giving once-off feedback to making a code contribution, 
to having long-term owners to help us drive changes. The Windows Operational 
Readiness is owned by the SIG Windows team as a sub-project. You can reach out 
to the team on the [Kubernetes Slack workspace](https://slack.k8s.io/) **#sig-windows** 

> channel. You can also explore the [Windows Operational Readiness test suite](https://raw.githubusercontent.com/kubernetes-sigs/windows-operational-readiness) 

Reviewer 2: This link doesn't seem right. How about https://github.com/kubernetes-sigs/windows-operational-readiness/#readme?
